src/kernels/rotary_embedding_kernels.cpp

```diff
@@ -0,0 +1,134 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// ============================================================================
+#include "rotary_embedding_kernels.h"
+
+namespace xft {
+
+void llamaSetCosSinCache(
+        const float *invFreq, float *embCos, float *embSin, int invFreqSize, int maxPositionEmbeddings, float scale) {
+
+#pragma omp parallel for
+    for (size_t i = 0; i < maxPositionEmbeddings; i++) {
+        float *pcos = embCos + i * invFreqSize;
+        float *psin = embSin + i * invFreqSize;
+
+        for (size_t j = 0; j < invFreqSize; j++) {
+            float tmp = i * invFreq[j];
+            float cosTmp = std::cos(tmp) * scale;
+            float sinTmp = std::sin(tmp) * scale;
+
+            pcos[j] = cosTmp;
+            psin[j] = sinTmp;
+        }
+    }
+}
+
+// def rotate_half(x):
+//     """Rotates half the hidden dims of the input."""
+//     x1 = x[..., : x.shape[-1] // 2]
+//     x2 = x[..., x.shape[-1] // 2 :]
+//     return torch.cat((-x2, x1), dim=-1)
+// def apply_rotary_pos_emb(q, k, cos, sin, position_ids):
+//     # The first two dimensions of cos and sin are always 1, so we can `squeeze` them.
+//     cos = cos.squeeze(1).squeeze(0)  # [seq_len, dim]
+//     sin = sin.squeeze(1).squeeze(0)  # [seq_len, dim]
+//     cos = cos[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+//     sin = sin[position_ids].unsqueeze(1)  # [bs, 1, seq_len, dim]
+//     q_embed = (q * cos) + (rotate_half(q) * sin)
+//     k_embed = (k * cos) + (rotate_half(k) * sin)
+//     return q_embed, k_embed
+//
+
+void llamaApplyRotaryPosEmbed(float *query, float *key, float *emb_cos, float *emb_sin, int qStride, int kStride,
+        int dim, int totSeqLen, int qHeads, int kHeads, const int *positionIds) {
+    const int half = (dim + 1) / 2;
+    const int heads = std::max(qHeads, kHeads);
+
+#pragma omp parallel for collapse(2)
```

Comment:
currently we have imbalance tasks inside omp block, we could optimize it in next stage.

Reply:
Each row of q and k can be parallel, so it should be balanced~

Reply:
when kHeads < qHeads, it may cause imbalance issue. but anyhow, next step.